pkg/controller/registry/containerregistry.go

```diff
@@ -0,0 +1,201 @@
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+// Connect sets up iam client using credentials from the provider
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, _, err := gcp.GetAuthInfo(ctx, c.client, mg)
+
+	if err != nil {
+		return nil, err
+	}
+
+	storageService, err := storage.NewService(ctx)
+
+	if err != nil {
+		return nil, errors.Wrap(err, errNewStorageClient)
+	}
+
+	return &external{client: c.client, storage: storageService, projectID: projectID}, nil
+}
+
+type external struct {
+	client    client.Client
+	storage   *storage.Service
+	projectID string
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ContainerRegistry)
+
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotGcr)
+	}
+
+	bucket, err := e.getBucket(e.getBucketName(cr))
+
+	if gcp.IsErrorNotFound(err) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetBucket)
+	}
+
+	populateCRFromBucket(cr, bucket)
+
+	cr.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  true,
+		ConnectionDetails: managed.ConnectionDetails{},
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.ContainerRegistry)
+
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotGcr)
+	}
+
+	location := gcrURL
+
+	if cr.Spec.ForProvider.Location != "" {
+		location = fmt.Sprintf("%s.%s", strings.ToLower(cr.Spec.ForProvider.Location), location)
+	}
+
+	ref, err := name.ParseReference(fmt.Sprintf("%s/%s/%s", location, e.projectID, testRepoName))
+
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	auth, err := authn.DefaultKeychain.Resolve(ref.Context().Registry)
+
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	if _, err := transport.NewWithContext(ctx, ref.Context().Registry, auth, http.DefaultTransport,
+		[]string{ref.Scope(transport.PushScope)}); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errHandshake)
+	}
+
+	return managed.ExternalCreation{}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	return nil
+}
+
+func populateCRFromBucket(cr *v1alpha1.ContainerRegistry, bucket *storage.Bucket) {
+	cr.Status.AtProvider.ID = bucket.Id
+	cr.Status.AtProvider.BucketLink = bucket.SelfLink
```

Comment:
nit: I would suggest doing this inline above instead of defining a function since it is just two lines.